contracts/vortex/CarbonVortex.sol

```diff
@@ -252,37 +858,109 @@ contract CarbonVortex is ICarbonVortex, Upgradeable, ReentrancyGuardUpgradeable,
         emit RewardsUpdated({ prevRewardsPPM: prevRewardsPPM, newRewardsPPM: newRewardsPPM });
     }
 
+    function _setPairDisabled(Token token) private {
+        bool prevPairStatus = _disabledPairs[token];
+
+        _disabledPairs[token] = !prevPairStatus;
+
+        emit PairDisabledStatusUpdated(token, prevPairStatus, !prevPairStatus);
+    }
+
     /**
-     * @dev set allowance to Bancor Network V3 to the max amount if it's less than the input amount
+     * @dev returns true if the auction price is below or equal to the minimum possible price
+     * @dev check if timeElapsed / priceDecayHalfLife >= 128
      */
-    function _setAllowance(Token token, uint256 inputAmount) private {
-        if (token.isNative()) {
-            return;
+    function _auctionPriceIsBelowMinimum(Token token) private view returns (bool) {
+        // cache trading start time to save gas
+        uint32 tradingStartTime = _tradingStartTimes[token];
+        // trading hasn't been enabled, return false
+        if (tradingStartTime == 0) {
+            return false;
         }
-        uint256 allowance = token.allowance(address(this), address(_bancorNetwork));
-        if (allowance < inputAmount) {
-            // increase allowance to the max amount if allowance < inputAmount
-            token.safeIncreaseAllowance(address(_bancorNetwork), type(uint256).max - allowance);
+        // get time elapsed since trading was enabled
+        uint32 timeElapsed = uint32(block.timestamp) - tradingStartTime;
+        // get the halflife for the token
+        uint32 currentPriceDecayHalfLife = token == _targetToken ? _targetTokenPriceDecayHalfLife : _priceDecayHalfLife;
+        // check if the maximum amount of halflifes have been reached
+        if (timeElapsed / currentPriceDecayHalfLife >= 128) {
+            return true;
+        } else {
+            return false;
         }
     }
 
-    function _validateTokens(Token[] calldata tokens) private view {
+    /**
+     * @dev returns the token amount available for trading
+     */
+    function _amountAvailableForTrading(Token token) private view returns (uint128) {
+        if (token == _targetToken) {
+            return _targetTokenSaleAmount.current;
+        } else {
+            return uint128(token.balanceOf(address(this)));
```

Comment:
Following the code in full (i.e., not just the diffs), I now think that the cast here is undesired to begin with.
Why leave the potential of reverting just because the balance is for some reason larger than 128 bits?
So instead, this function can just return `uint256`.
It is used only for checks, i.e., not for actual assignment of values or anything like that.
So applying this should not cause any issue other than a bit of a headache in casting the other operand in each of these checks to `uint256`.

Reply:
added safeCast
dealing only with uint128 max values for token balances and amounts in the entire code. Doesn't make sense to support balances > uint128